File: shoebill_plugins/JobPlugin/src/main/java/lt/ltrp/AbstractFaction.java
package lt.ltrp;

import lt.ltrp.object.Faction;
import lt.ltrp.object.FactionRank;
import lt.ltrp.object.JobVehicle;
import lt.ltrp.object.Rank;
import net.gtaun.shoebill.data.Location;
import net.gtaun.util.event.EventManager;

import java.util.ArrayList;
import java.util.Collection;
import java.util.stream.Collectors;

/**
 * @author Bebras
 *         2015.12.06.
 */
public abstract class AbstractFaction extends AbstractJob implements Faction {

    public static final int INVALID_ID = 0;

    private Collection<FactionRank> ranks;
    private Collection<Integer> leaderIds;
    private Collection<JobVehicle> vehicles;
    private int budget;

    @JobProperty("is_chat_enabled")
    private boolean isChatEnabled;

    public AbstractFaction(int id, String name, Location location, int basePaycheck, EventManager eventManager) {
        super(id, name, location, basePaycheck, eventManager);
        this.isChatEnabled  = true;
    }

    public AbstractFaction(EventManager eventManager) {
        super(eventManager);
        this.isChatEnabled = true;
    }


    @Override
    public Collection<JobVehicle> getVehicles() {
        return vehicles;
    }

    @Override
    public void setVehicles(Collection<JobVehicle> vehicles) {
        this.vehicles = vehicles;
    }

    @Override
    public Collection<JobVehicle> getVehicles(Rank rank) {
        return rank instanceof FactionRank ? getVehicles((FactionRank)rank) : new ArrayList<>();
    }

    @Override
    public Collection<JobVehicle> getVehicles(FactionRank rank) {
        return vehicles == null ?
                new ArrayList<>(0) :
                vehicles.stream().filter(v -> v.getRequiredRank().equals(rank)).collect(Collectors.toList());
    }

    public void addLeader(int userid) {
        leaderIds.add(userid);
    }

    public Collection<Integer> getLeaders() {
        return leaderIds;
    }

    public void removeLeader(int userid) {
        leaderIds.remove(userid);
    }

    @Override
    public int getBudget() {
        return budget;
    }

    @Override
    public void setBudget(int budget) {
        this.budget = budget;
    }

    @Override
    public void addVehicle(JobVehicle vehicle) {
        vehicles.add(vehicle);
    }


    @Override
    public Collection<FactionRank> getRanks() {
        return ranks;
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof Faction && ((Faction) o).getUUID() == this.getUUID();
    }

    @Override
    public FactionRank getRank(int id) {
        for(FactionRank rank : ranks) {
            if(rank.getNumber() == id) {
                return rank;
            }
        }
        return null;
    }

    @Override
    public void setRanks(Collection<? extends Rank> collection) {
        if(ranks == null) {
            ranks = new ArrayList<>();
        }
        collection.stream().filter(r -> r instanceof FactionRank).forEach(r -> ranks.add((FactionRank)r));
    }

    @Override
    public String toString() {
        return super.toString() + String.format("Faction. Id %s Name:%s Rank count:%d vehicle count:%d", getUUID(), getName(), getRanks().size(), vehicles.size());
    }

    @Override
    public boolean isChatEnabled() {
        return isChatEnabled;
    }

    @Override
    public void setChatEnabled(boolean chat) {
        this.isChatEnabled = chat;
    }
}

File: shoebill_plugins/JobPlugin/src/main/java/lt/ltrp/object/impl/AbstractFaction.kt
package lt.ltrp.`object`.impl

import lt.ltrp.`object`.PlayerData
import lt.ltrp.constant.JobProperty
import lt.ltrp.job.JobController
import lt.ltrp.job.`object`.Faction
import lt.ltrp.job.`object`.FactionRank
import lt.ltrp.job.`object`.JobVehicle
import net.gtaun.util.event.EventManager


/**
 * @author Bebras
 *         2015.12.06.
 */
abstract class AbstractFaction(uuid: Int,
                               override val ranks: MutableList<out FactionRank>,
                               override var budget: Int,
                               @JobProperty("is_chat_enabled") override var isChatEnabled: Boolean,
                               private var leaderCollection: MutableCollection<PlayerData>,
                      eventManager: EventManager):
        AbstractJob(uuid, eventManager), Faction {

    constructor(uuid: Int, eventManager: EventManager):
        this(uuid, mutableListOf<FactionRank>(), 0, false, mutableSetOf<PlayerData>(), eventManager) {

    }

    override fun getVehicles(rank: FactionRank): kotlin.collections.Collection<JobVehicle> {
        return getVehicles(rank)
    }

    override fun getRankByNumber(number: Int): FactionRank? {
        return ranks.firstOrNull { it.number == number }
    }

    override fun getRankByUUID(uuid: Int): FactionRank? {
        return ranks.firstOrNull { it.UUID == uuid }
    }

    override fun getLeaders(): Collection<PlayerData> {
        return leaderCollection
    }

    override fun addLeader(playerData: PlayerData) {
        if(!leaderCollection.contains(playerData)) {
            leaderCollection.add(playerData)
            JobController.instance.addLeader(this, playerData)
        }
    }

    override fun removeLeader(playerData: PlayerData) {
        if(leaderCollection.contains(playerData)) {
            leaderCollection.remove(playerData)
            JobController.instance.removeLeader(this, playerData)
        }
    }

    override fun equals(other: Any?): Boolean {
        return super.equals(other) && other is Faction && other.UUID == this.UUID
    }

    override fun toString(): String {
        return super.toString() +
                String.format("Faction. Id %s Name:%s JobRank count:%d vehicle count:%d",
                        UUID, name, ranks.size, vehicles.size)
    }

    override fun hashCode(): Int {
        var result = ranks.hashCode()
        result = 31 * result + budget
        result = 31 * result + isChatEnabled.hashCode()
        result = 31 * result + leaderCollection.hashCode()
        return result
    }

}
